### PR TITLE
log: improvements to sparse macros, fancy logger.

### PR DESCRIPTION
**Pull request by htuch**
Inspired by a log statement needed for #10815, this patch:

* Adds _MISC variants of sparse macros.

* Sends _MISC log messages to fancy logger (these were previously
  ignored). This is a bit weird, as "misc" is also a logger and
  we use file-specific loggers in fancy logger, but in reality, most
  folks treat this macro as the generic Envoy log statement.

* Avoid evaluating expressions in ENVOY_LOG* with fancy logger when
  beneath the configured log level.

Risk level: Low
Testing: Mostly covered by log_macros_test already, added some _MISC
  coverage in SparseLogMacrosTest.

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

File: source/common/common/fancy_logger.cc [hunk 1]
```diff
@@ -90,6 +90,15 @@ void FancyContext::setAllFancyLoggers(spdlog::level::level_enum level)
   }
 }
 
+FancyLogLevelMap FancyContext::getAllFancyLogLevels() ABSL_LOCKS_EXCLUDED(fancy_log_lock_) {
```
Comment (oschaaf) [inline source/common/common/fancy_logger.cc:93]: I lack some context here, but if this is a production code utility, would this deserve a unit test of its own, and maybe add caching to it? 
If not, would it be worth considering moving this into test/ as right now it's only used there? 
I don't want to bloat this PR, just curious.

**Review by oschaaf — COMMENTED** [review]
a nit and a question, otherwise looks good to me.

File: test/common/common/log_macros_test.cc [hunk 2]
```diff
@@ -142,69 +142,124 @@ void spamCall(std::function<void()>&& call_to_spam, const uint32_t num_threads)
...
+    if (use_misc_macros_) {
+      ENVOY_LOG_PERIODIC_MISC(error, 1s, "foo7 '{}'", evaluations()++);
+    } else {
+      ENVOY_LOG_PERIODIC(error, 1s, "foo7 '{}'", evaluations()++);
+    }
+  }
+  std::atomic<int32_t>& evaluations() { MUTABLE_CONSTRUCT_ON_FIRST_USE(std::atomic<int32_t>); };
+
+  const bool use_misc_macros_;
+  LogLevelSetter save_levels_{spdlog::level::info};
```
[the remark this answers is not in the document]
Comment (htuch) [reply test/common/common/log_macros_test.cc:208]: It's RAII, it saves all the log levels when object is created and restores them when destroyed.

File: source/common/common/fancy_logger.cc (hunk 1 above)
Comment (htuch) [reply source/common/common/fancy_logger.cc:93]: fancy_logger.h comments say this is test only, I don't mind renaming to `getAllFancyLogLevelsForTest()` if that makes it clearer. Not sure how to move it out of this file as it needs access to internal fields.

**Review by oschaaf — DISMISSED** [review]
Thanks for the explanations. LGTM

File: source/common/common/fancy_logger.cc (hunk 1 above)
Comment (ggreenway) [reply source/common/common/fancy_logger.cc:93]: Yes, please name test-only code as such. Makes it much easier to understand.

**Review by ggreenway — CHANGES_REQUESTED** [review]
LGTM, modulo renaming `....ForTest`

**Review by ggreenway — APPROVED** [review]

**Pull request merged**